src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+    hashes: List[CycloneDxHashObject] = attr.ib(factory=list)
+
+
+class CycloneDxComponentType(str, Enum):
+    APPLICATION = "application"
+    FRAMEWORK = "framework"
+    LIBRARY = "library"
+    CONTAINER = "container"
+    OPERATING_SYSTEM = "operating-system"
+    DEVICE = "device"
+    FIRMWARE = "firmware"
+    FILE = "file"
+
+
+class CycloneDxComponentScope(str, Enum):
+    REQUIRED = "required"
+    OPTIONAL = "optional"
+    EXCLUDED = "excluded"
+
+
+@attr.s
+class CycloneDxComponent():
+    name: str = attr.ib()
+    version: str = attr.ib()
+    bom_ref: str = attr.ib(default=None)
+    group: str = attr.ib(default=None)
+    copyright: str = attr.ib(default=None)
+    author: str = attr.ib(default=None)
+    description: str = attr.ib(default=None)
+    purl: str = attr.ib(default=None)
+    hashes: List[CycloneDxHashObject] = attr.ib(factory=list)
+    licenses: List[CycloneDxLicenseEntry] = attr.ib(factory=list)
+    externalReferences: List[CycloneDxExternalRef] = attr.ib(factory=list)
+    type: CycloneDxComponentType = attr.ib(
+        default=CycloneDxComponentType.LIBRARY)
+    properties: List[CycloneDxAttribute] = attr.ib(factory=list)
+    scope: CycloneDxComponentScope = attr.ib(default=CycloneDxComponentScope.REQUIRED)
+
+
+@attr.s
+class CycloneDxDependency:
+    ref: str = attr.ib()
+    dependsOn: str = attr.ib(factory=list)
+
+
+@attr.s
+class CycloneDxBom():
+    bomFormat: str = attr.ib(init=False, default="CycloneDX")
+    specVersion: str = attr.ib(init=False, default="1.3")
+    serialNumber: str = attr.ib(
+        init=False, default="urn:uuid:" + str(uuid.uuid4()))
+    version: int = attr.ib(init=False, default=1)
+    metadata: CycloneDxMetadata = attr.ib(default=CycloneDxMetadata())
+    components: List[CycloneDxComponent] = attr.ib(factory=list)
+    dependencies: List[CycloneDxDependency] = attr.ib(factory=list)
+
+
+def get_tool_header(version: str) -> dict:
+    return {
+        "vendor": "nexB Inc.",
+        "name": "scancode-toolkit",
+        "version": version
+    }
+
+
+def get_external_ref_from_key(package: dict, key: str) -> CycloneDxExternalRef:
+    url = package.get(key)
+    type = url_type_mapping[key]
+    if url is not None and type is not None:
+        return CycloneDxExternalRef(url=url, type=type)
+    return None
+
+
+def get_license_entry_from_license_expression(license_expression: str) \
+        -> CycloneDxLicenseEntry:
+    """query our list of known licenses to see if we can resolve
+    the license_expression to a single entry"""
+
+    # exit early if we don't have a valid license_expression
+    if license_expression is None:
+        return None
+
+    license = known_licenses.get(license_expression)
+    if license is not None:
+        # attempt to set either official OSI URL or any license text URL
+        url = license.osi_url
+        if license.text_urls:
+            url = license.text_urls[0]
+        lic = CycloneDxLicense(id=license.spdx_license_key,
```

Comment:
Do you need an SPDX license expression? if so we can get this directly instead.

What is CycloneDxLicenseEntry? a single license or an expression?